rust/src/lib.rs

```diff
@@ -90,6 +91,34 @@ pub mod http {
         Some((split.next()?, split.next()?))
     }
 
+    fn decompress<R: Read>(reader: &mut BufReader<R>, encoding: ContentEncoding) -> Vec<u8> {
```

Comment:
if we don't care about performance but care about the readability of code, having a 'decompressed reader' might be simple and more comprehensible.

```rust
fn decompressor<R: Read, D: Read>(reader: &mut BufReader<R>, encoding: ContentEncoding) -> Box<dyn D> {
    match encoding {
        ContentEncoding::Gzip => Box::new(GzDecoder::new(reader)),
        ContentEncoding::Deflate => Box::new(DeflateDecoder::new(reader)),
        ContentEncoding::Identity => Box::new(reader),
   }
}
```

Then, `decompress(reader, encoding).read_to_end(&mut body)` will work.